File: 75 days Challenge Leetcode/HashMap+Set/Day-22/Solution.py
from collections import Counter

class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        # Check if the words are of different lengths
        if len(word1) != len(word2):
            return False

        # Sets to store unique characters in each word
        set_word1 = set(word1)
        set_word2 = set(word2)

        # If the sets of characters are not the same, words can't be "close"
        if set_word1 != set_word2:
            return False

        # Count the frequency of each character in both words
        freq_word1 = Counter(word1)
        freq_word2 = Counter(word2)

        # Get frequency values and sort them
        freq_values1 = sorted(freq_word1.values())
        freq_values2 = sorted(freq_word2.values())

        # Compare the sorted frequency lists
        return freq_values1 == freq_values2

File: 75 days Challenge Leetcode/HashMap+Set/Day-22/Solution.java
import java.util.*;

class Solution {
    public boolean closeStrings(String word1, String word2) {
        
        if (word1.length() != word2.length()) return false;

        // Sets to store unique characters in each word
        Set<Character> setWord1 = new HashSet<>();
        Set<Character> setWord2 = new HashSet<>();

        // Maps to store the frequency of each character in each word
        Map<Character, Integer> freqWord1 = new HashMap<>();
        Map<Character, Integer> freqWord2 = new HashMap<>();

        // Populate set and frequency map for word1
        for (char c : word1.toCharArray()) {
            setWord1.add(c);           // Add character to set
            freqWord1.put(c, freqWord1.getOrDefault(c, 0) + 1); // Increment frequency
        }

        // Populate set and frequency map for word2
        for (char c : word2.toCharArray()) {
            setWord2.add(c);           // Add character to set
            freqWord2.put(c, freqWord2.getOrDefault(c, 0) + 1); // Increment frequency
        }

        // If the sets of characters are not the same, words can't be "close"
        if (!setWord1.equals(setWord2)) return false;

        // Lists to store the frequency counts of each character
        List<Integer> freqValues1 = new ArrayList<>(freqWord1.values());
        List<Integer> freqValues2 = new ArrayList<>(freqWord2.values());

        // Sort both frequency lists to compare them
        Collections.sort(freqValues1);
        Collections.sort(freqValues2);

        // Compare the frequency distributions
        return freqValues1.equals(freqValues2);
    }
}
